### PR TITLE
Badging system

### LINKED ISSUE #18: Badging system
### What is the expected behavior?

Implement badging system for members of ecosystem.

Badging should be dynamic and reflect state of json db with members. 

Potential solution: https://github.com/marketplace/actions/dynamic-badges

Readme should have collapsable section for each of member of ecosystem which has code of badge, so developers can add this code in their readme to showcase status of membership.

Later on we will implement CI to submit badging for members repositories. 

**Related stories**: as core team member/curator I want badging system to be automated so that badges dynamically checks repository membership (ex: gray if checks did not passed or purple of checks passed)

**Epic**: #9

### PR DESCRIPTION
**Pull request by mickahell**
### Summary
Create dynamic badge for member of the ecosystem.
Test run : https://github.com/mickahell/ecosystem/runs/4115743386?check_suite_focus=true
MR from run : https://github.com/mickahell/ecosystem/pull/7

#### WF
- [x] Creating external action for badging 
- [x] Create update badge wf
- [x] Link badge action on submission success

#### Manager
- [x] Update manager `dispatch_check_workflow` function
- [x] Create manager  `update_badges` function

#### Tests
- [x] Add unit tests
- [x] Tests wf

#### Doc
- [x] Update Readme

---
Closes #18 
Epic #9

### CODE REVIEW

**Comment by mickahell** [comment]
Need #82 before to avoid conflict

**Comment by mickahell** [comment]
@IceKhan13 what you thing of that :) ?

**Comment by mickahell** [comment]
Need #85 before to avoid conflict

**Review by IceKhan13 — CHANGES_REQUESTED** [review]
Wow! Cool idea about badges to save them as files. It might work :) 

I think we need to move badge check as separate workflow. Something like:
- get all repos
- for each repo check if latest tests are passed or not

It will allow us to update badges independently of tests. So, we basically separating different functionalities. Loose coupling is good practice :)

**Comment by mickahell** [comment]
> Wow! Cool idea about badges to save them as files. It might work :)
> 
> I think we need to move badge generation as separate workflow. Something like:
> 
> * get all repos
> * for each repo check if latest tests are passed or not
> 
> It will allow us to update badges independently of tests. So, we basically separating different functionalities. Loose coupling is good practice :)

Hmm, so basically the same as the weekly check wf.
But for the first creating badge, don't you want to keep it in the submission wf and then having another wf to check version and generate fix badge ? 

In fact, keeping the badges job in the submission and weekly check wf AND having a wf for updating badge ?
So 2 for the automation part and the last one for manual update. What do you think ? @IceKhan13

**Comment by mickahell** [comment]
@IceKhan13 I externalized the badging job to an external action, call it in the submission and weekly-check and create a new wf to update badges independently.

What do you think of that :)

**Comment by mickahell** [comment]
I have a strange bug, the job `Get project for check` don't want to set output my tests var : https://github.com/mickahell/ecosystem/runs/4190787950?check_suite_focus=true

It's working in local but not with action :/

**Comment by mickahell** [comment]
never mind, I find why, in the batch wf we set to pull the branch main and so my code are on another branch... th's why ^^'

So @IceKhan13 have fun ;) (I'll just correct the black now)

**Comment by mickahell** [comment]
I just remember to clean a job, whi isn't necessary for badge wf

**Review by IceKhan13 — CHANGES_REQUESTED** [review]
Looks good, but don't you think it is overcompicated a little bit?

We have separate workflow for each project for tests because tests are REALLY HEAVY and they need to be executed in parallel, but for badges we do not need this :) 

I think something super simple like
```python
from urllib.request import urlretrieve

class Manager:
    ...
    
    def update_badges(self):
         """Updates badges for projects."""
         badges_folder_path = ...
         community_projects = self.dao.get_projects_by_tier(COMMUNITY)
         for project in community_projects:
             image_url = f"https://img.shields.io/static/v1?label=Qiskit&message=${{ project.repo_name }}&color=grey"
             target_image_name = f"{badges_folder_path}/project.repo_url.svg"
             urlretrieve(image_url, image_local_path)
```

and then single workflow will look simple as that:

```yaml
steps:
- name: Update badges
   runs: python manager.py update_badges
- name: Create PR with updates badges
  ...
```

I think it will simplify a workflow and tests a lot

What do you think?

**Comment by mickahell** [comment]
hahah Indeed ^^' 
I'll fix it tonight 

I didn't know this way of doing : `color = "blueviolet" is tests_passed else "gray"` with the `is`

**Comment by IceKhan13** [comment]
> I didn't know this way of doing : `color = "blueviolet" is tests_passed else "gray"` with the `is`

it suppose to be `if` 😄

**Comment by mickahell** [comment]
@IceKhan13 done :) and it's working --> here the [test](https://github.com/mickahell/ecosystem/pull/17)

File: ecosystem/manager.py [hunk 1]
```diff
@@ -80,13 +84,41 @@ def dispatch_check_workflow(
...
+        """Updates badges for projects."""
+        current_dir = os.path.dirname(os.path.abspath(__file__))
+        badges_folder_path = "{}/../badges".format(current_dir)
+
+        tests = []
+        community_projects = [
+            repo.name for repo in self.dao.get_repos_by_tier("COMMUNITY")
+        ]
+        tests_list = [
+            repo.tests_results for repo in self.dao.get_repos_by_tier("COMMUNITY")
```
Comment (IceKhan13) [inline ecosystem/manager.py:103]: you can reduce number of DB calls to one. No need to call `get_repos_by_tier` 2 times.

File: ecosystem/manager.py [hunk 2]
```diff
@@ -80,13 +84,41 @@ def dispatch_check_workflow(
...
+        ]
+
+        for i in tests_list:
```
Comment (IceKhan13) [inline ecosystem/manager.py:106]: `i` is not a good name for variable that represents test result 😄 
Usually `i` is for some kind of index.

File: ecosystem/manager.py [hunk 3]
```diff
@@ -80,13 +84,41 @@ def dispatch_check_workflow(
...
+            test_result = [str(repo.passed) for repo in i]
+            if "False" not in test_result:
+                tests.append(True)
+            else:
+                tests.append(False)
+
+        for project, tests_out in zip(community_projects, tests):
```
Comment (IceKhan13) [inline ecosystem/manager.py:113]: as I mentioned above we do not need to call db 2 times. We can just do something like:
```python
for project in self.dao.get_repos_by_tier(tier):
    tests_passed = all([result.passed for result in project.test_results])
```

File: .github/workflows/submission.yml [hunk 4]
```diff
@@ -114,7 +114,14 @@ jobs:
                                        --repo_affiliations="${{ steps.parse-issue.outputs.SUBMISSION_AFFILIATIONS }}" \
                                        --repo_labels="${{ steps.parse-issue.outputs.SUBMISSION_LABELS }}"
         python manager.py generate_readme
-
+    - name: Generate badge
```
Comment (IceKhan13) [inline .github/workflows/submission.yml:117]: Do we still need this?

File: .github/workflows/project-check.yml [hunk 5]
```diff
@@ -111,6 +103,24 @@ jobs:
       run: |
         echo "Pass log is ${{ env.PASS_LOG }}"
 
+    - name: Generate badge
```
Comment (IceKhan13) [inline .github/workflows/project-check.yml:106]: As we discussed workflow for badges should not be part of project checks. We can trigger it after we merge PR for project checks

File: .github/actions/update-badges/action.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+name: 'Create badge'
```
Comment (IceKhan13) [inline .github/actions/update-badges/action.yml:1]: Do we still need this?

**Review by IceKhan13 — CHANGES_REQUESTED** [review]

**Comment by IceKhan13** [comment]
Almost there!

File: ecosystem/manager.py [hunk 7]
```diff
@@ -80,13 +84,41 @@ def dispatch_check_workflow(
...
+            tests_passed = tests_out
+            color = "blueviolet" if tests_passed else "gray"
+            url = f"https://img.shields.io/static/v1?label=Qiskit&message={project}&color={color}"
```
Comment (IceKhan13) [inline ecosystem/manager.py:116]: also lets do `https://img.shields.io/static/v1?label=Qiskit&message=Ecosystem&color={color}`

Comment (mickahell) [reply ecosystem/manager.py:116]: If we put Ecosystem, means we can't control someone to put badge on it's repo even if its project doesn't it inside our db

File: .github/workflows/submission.yml (hunk 4 above)
Comment (mickahell) [reply .github/workflows/submission.yml:117]: For the automation yes

File: .github/workflows/project-check.yml (hunk 5 above)
Comment (mickahell) [reply .github/workflows/project-check.yml:106]: I waned to keep it for the automation

File: .github/actions/update-badges/action.yml (hunk 6 above)
Comment (mickahell) [reply .github/actions/update-badges/action.yml:1]: Depends, but yes we can just put a simple job

File: ecosystem/manager.py (hunk 2 above)
Comment (mickahell) [reply ecosystem/manager.py]: Ah,  I like using i in my for loop as it's only local 😅

**Comment by mickahell** [comment]
@IceKhan13 ready :)

File: tests/daos/test_dao.py [hunk 8]
```diff
@@ -117,3 +144,42 @@ def test_add_test_result(self):
                 TestResult(True, "0.18.2", TestType.DEV_COMPATIBLE),
             ],
         )
+
+    def test_update_badges(self):
```
Comment (IceKhan13) [inline tests/daos/test_dao.py:148]: Can we move tests for badge to `test_manager` file or other appropriate place 😄 
It should not live in database tests.

**Review by IceKhan13 — CHANGES_REQUESTED** [review]
Everything is good and I've tested batch badges! Once tests are moved to the right place we can merge

File: tests/daos/test_dao.py (hunk 8 above)
Comment (IceKhan13) [reply tests/daos/test_dao.py]: Nevermind, I did it :)

**Review by IceKhan13 — APPROVED** [review]

**Pull request merged**